iroh-metrics/src/lib.rs

```diff
@@ -141,3 +153,97 @@ pub fn get_current_trace_id() -> TraceId {
         .span_context()
         .trace_id()
 }
+
+#[derive(PartialEq)]
+pub enum Collector {
+    #[cfg(feature = "gateway")]
+    Gateway,
+    #[cfg(feature = "resolver")]
+    Resolver,
+    #[cfg(feature = "bitswap")]
+    Bitswap,
+    #[cfg(feature = "store")]
+    Store,
+}
+
+#[allow(unused_variables, unreachable_patterns)]
+pub async fn record<M>(c: Collector, m: M, v: u64)
+where
+    M: MetricType + std::fmt::Display,
+{
+    let r = CORE.read().await;
+    if r.enabled() {
+        match c {
+            #[cfg(feature = "gateway")]
+            Collector::Gateway => r.gateway_metrics().record(m, v),
+            #[cfg(feature = "resolver")]
+            Collector::Resolver => r.resolver_metrics().record(m, v),
+            #[cfg(feature = "bitswap")]
+            Collector::Bitswap => r.bitswap_metrics().record(m, v),
+            #[cfg(feature = "store")]
+            Collector::Store => r.store_metrics().record(m, v),
+            _ => panic!("not enabled/implemented"),
+        };
+    }
+}
+
+#[allow(unused_variables, unreachable_patterns)]
+pub fn record_sync<M>(c: Collector, m: M, v: u64)
+where
+    M: 'static + MetricType + std::fmt::Display + Send,
+{
+    tokio::spawn(async move {
+        record(c, m, v).await;
+    });
+}
+
+#[allow(unused_variables, unreachable_patterns)]
+pub async fn observe<M>(c: Collector, m: M, v: f64)
+where
+    M: HistogramType + std::fmt::Display,
+{
+    let r = CORE.read().await;
+    if r.enabled() {
+        match c {
+            #[cfg(feature = "gateway")]
+            Collector::Gateway => r.gateway_metrics().observe(m, v),
+            #[cfg(feature = "resolver")]
+            Collector::Resolver => r.resolver_metrics().observe(m, v),
+            #[cfg(feature = "bitswap")]
+            Collector::Bitswap => r.bitswap_metrics().observe(m, v),
+            #[cfg(feature = "store")]
+            Collector::Store => r.store_metrics().observe(m, v),
+            _ => panic!("not enabled/implemented"),
+        };
+    }
+}
+
+#[allow(unused_variables, unreachable_patterns)]
+pub fn observe_sync<M>(c: Collector, m: M, v: f64)
+where
+    M: 'static + HistogramType + std::fmt::Display + Send,
+{
+    tokio::spawn(async move {
+        observe(c, m, v).await;
+    });
+}
+
+async fn print() {
+    let r = CORE.read().await;
+    println!("{}", String::from_utf8(r.encode()).unwrap());
+}
+
+async fn enable() {
+    let mut r = CORE.write().await;
+    r.set_enabled(true);
+}
+
+async fn disable() {
+    let mut r = CORE.write().await;
+    r.set_enabled(false);
+}
+
+pub async fn p2p_metrics() -> p2p::Metrics {
```

Comment:
p2p metrics are sourced from libp2p so without re-doing the whole thing (and assuming maintenance of it) we just use a pass through type. It also changes the metrics dynamics for p2p as libp2p works on an event based platform.